File: 0118-pascals-triangle/0118-pascals-triangle.cpp
class Solution {
public:
    vector<int> generateNextRow(const vector<int>* const prevVec) {
        vector<int> v(0);
        v.push_back(1);
        for (int i = 0; i < prevVec->size() - 1; ++i) {
            v.push_back((*prevVec)[i] + (*prevVec)[i + 1]);
        }
        v.push_back(1);
        return v;
    }
    vector<vector<int>> generate(int numRows) {
        if (numRows == 0) {
            return {};
        }
        vector<vector<int>> res(numRows);
        res[0] = vector<int>{ 1 };
        for (int i = 1; i < numRows; ++i) {
            res[i] = generateNextRow(&res[i - 1]);
        }
        return res;
    }
};

File: 0118-pascals-triangle/0118-pascals-triangle.rs
impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        if num_rows == 1 {
            return vec![vec![1]];
        } else if num_rows == 2 {
            return vec![vec![1], vec![1, 1]];
        }
        let mut res = vec![vec![1], vec![1,1]];
        for i in 2..num_rows {
            let mut cur_vec = vec![1];
            for j in 1..i {
                cur_vec.push(res[i as usize - 1][j as usize] + res[i as usize - 1][j as usize - 1]);
            }
            cur_vec.push(1);
            res.push(cur_vec);
        }
        res
    }
}